### PR TITLE
introduce voting bond weight factor

### LINKED ISSUE #7612: Charge VotingBond in elections-phragmen for weight as well
with #7040, the module will track the amount of deposit for all voters. Moreover, the voting bond is solely based on the amout of storage that is kept on chain. 

I think this is not enough. Each voter will will impose a lot of additional computation on the next election term. They should, somehow at least reserve some money for this, ensuring that it is even harder to DoS the chain with this. 

I propose expanding the voting bond deposit to 

```
VotingBondStorageBase
VotingBondStorageFactor
VotingBondWeightFactor
```

The weight associated with the last one will essentially be: 

```
// given voter with `num_votes` votes being casted. 
let voter_weight = T::WeightInfo::election_phragmen(1, num_votes, 0)
// ^^ basically gives the additional weight added for 1 votes, and `num_votes` votes being casted.
let weight_deposit = voter_weight * T::VotingBondWeightFactor::get()
``

### PR DESCRIPTION
**Pull request by ropottnik**
closes #7612 
* add voting bond weight factor
  * I set this quite randomly to `1 MILLICENTS`. This will most likely need to be adjusted but I had no intuition on what was appropriate
* rename VotingBondBase to VotingBondStorageBase
* rename VotingBondFactor to VotingBondStorageFactor

Polkadot companion PR: https://github.com/paritytech/polkadot/pull/2316

### CODE REVIEW

**Comment by ropottnik** [comment]
> Thanks for the PR! are you done? The PR does not yet do the second part. Namely, you are not really using the new weight factor anywhere 😄. You should compute the additional weight that each voter will impose:
> 
> ```
> // given voter with `num_votes` votes being casted. 
> let voter_weight = T::WeightInfo::election_phragmen(1, num_votes, 0)
> ```
> 
> And then compute the deposit for it:
> 
> ```
> // ^^ basically gives the additional weight added for 1 vote1, and `num_votes` votes being casted.
> let weight_deposit = voter_weight * T::VotingBondWeightFactor::get()
> ```
> 
> and the reserve this additionally in the `vote` extrinsic.

Oh, I thought I had done that at `frame/elections-phragmen/src/lib.rs#665` in the diff. Is it not right like this?

**Comment by kianenigma** [comment]
ahh damn, I didn't look carefully! I'll review soon.

File: frame/elections-phragmen/src/lib.rs [hunk 1]
```diff
@@ -1775,6 +1790,49 @@ mod tests {
 		});
 	}
 
+	#[test]
+	fn voting_reserves_bond_per_vote_for_computational_effort() {
+		ExtBuilder::default().balance_factor(1000000000).voter_bond_weight_factor(1).build_and_execute(|| {
```
Comment (kianenigma) [inline frame/elections-phragmen/src/lib.rs:1795]: these tests will break every time we re-benchmark the pallet. Better is to make the `type WeightInfo` mock-able. You can use this as an inspiration:

**Review by kianenigma — COMMENTED** [review]
Soorty for my delay! 

Looks good! just needs to fix the test and a quick audit.

**Comment by ropottnik** [comment]
thanks for your feedback @kianenigma 
I have now added a `MockWeightInfo` struct for testing the `VotingBondWeightFactor`. However, I didn't add the case distinction from the PR you referenced because it would have required adding a `MockWeightInfo` `boolean` type in the pallet `Config` trait, I believe and therefore would require unnecessary boilerplate on consuming runtimes. Also I was worried of circular dependencies when defining a boolean type in the `lib.rs` then use that property in `weights.rs` and then again using its results in the tests of `lib.rs`. please correct me if I my thoughts are incorrect, I want to learn ;-)

File: frame/elections-phragmen/src/lib.rs [hunk 2]
```diff
@@ -654,8 +658,12 @@ decl_module! {
 impl<T: Config> Module<T> {
 	/// The deposit value of `count` votes.
 	fn deposit_of(count: usize) -> BalanceOf<T> {
-		T::VotingBondBase::get().saturating_add(
-			T::VotingBondFactor::get().saturating_mul((count as u32).into())
+		T::VotingBondStorageBase::get().saturating_add(
+			T::VotingBondStorageFactor::get().saturating_mul((count as u32).into())
+		).saturating_add(
+			T::VotingBondWeightFactor::get().saturating_mul(
+				(T::WeightInfo::election_phragmen(1, (count as u32).into(), 0) as u32).into()
```
Comment (kianenigma) [inline frame/elections-phragmen/src/lib.rs:665]: Here you should not convert weight (u64) to u32 and then `.into()` it to balance. This can easily loose accuracy. 

Instead, you should directly convert weight to balance with `saturated_into`, see `SaturatedConversion` trait for reference.

File: frame/elections-phragmen/src/weights.rs [hunk 3]
```diff
@@ -228,3 +228,86 @@ impl WeightInfo for () {
...
 }
+
+pub struct MockWeightInfo<T>(PhantomData<T>);
+impl<T: frame_system::Config> WeightInfo for MockWeightInfo<T> {
+	fn vote_equal(v: u32, ) -> Weight {
+		(45_157_000 as Weight)
+			// Standard Error: 6_000
+			.saturating_add((399_000 as Weight).saturating_mul(v as Weight))
+			.saturating_add(RocksDbWeight::get().reads(5 as Weight))
+			.saturating_add(RocksDbWeight::get().writes(2 as Weight))
```
Comment (kianenigma) [inline frame/elections-phragmen/src/weights.rs:239]:
```suggestion
		(45_157_000 as Weight)
			unreachable!();
```

Same should be applied to all the ones that are not used.

File: frame/elections-phragmen/src/weights.rs [hunk 4]
```diff
@@ -228,3 +228,86 @@ impl WeightInfo for () {
...
+			.saturating_add((114_815_000 as Weight).saturating_mul(v as Weight))
+			// Standard Error: 53_000
+			.saturating_add((49_000 as Weight).saturating_mul(d as Weight))
+			.saturating_add(RocksDbWeight::get().reads(4 as Weight))
+			.saturating_add(RocksDbWeight::get().reads((3 as Weight).saturating_mul(v as Weight)))
+			.saturating_add(RocksDbWeight::get().writes((3 as Weight).saturating_mul(v as Weight)))
+	}
+	fn election_phragmen(c: u32, v: u32, e: u32, ) -> Weight {
+		(0 as Weight)
+			// Standard Error: 1_940_000
```
Comment (kianenigma) [inline frame/elections-phragmen/src/weights.rs:305]:
```suggestion
```

File: frame/elections-phragmen/src/weights.rs [hunk 5]
```diff
@@ -228,3 +228,86 @@ impl WeightInfo for () {
...
+			.saturating_add((1 as Weight).saturating_mul(c as Weight))
+			// Standard Error: 807_000
```
Comment (kianenigma) [inline frame/elections-phragmen/src/weights.rs:307]:
```suggestion
```

File: frame/elections-phragmen/src/weights.rs [hunk 6]
```diff
@@ -228,3 +228,86 @@ impl WeightInfo for () {
...
+			.saturating_add((1 as Weight).saturating_mul(v as Weight))
+			// Standard Error: 55_000
```
Comment (kianenigma) [inline frame/elections-phragmen/src/weights.rs:309]:
```suggestion
```

**Review by kianenigma — COMMENTED** [review]
Yes no need to add a config to all pallets. 

Please see my note about conversion, and one last thing: let's move `MockWeightInfo` to `mock.rs`, because `weight.rs` will be updated by bots.

**Comment by ropottnik** [comment]
from my point of view I have applied your suggestions. can you check again? 

thanks!

File: frame/elections-phragmen/src/mock.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
+use frame_support::{dispatch::Weight, weights::constants::RocksDbWeight};
```
Comment (kianenigma) [inline frame/elections-phragmen/src/mock.rs:1]: ahh sorry I thought this pallet had a mock file 🤦, I meant inside the rest of the mock stuff (inside `#[cfg(test)] mod test {...}`) but it is okay.

**Review by kianenigma — APPROVED** [review]
Thanks!

**Comment by kianenigma** [comment]
> However, I didn't add the case distinction from the PR you referenced because it would have required adding a MockWeightInfo boolean type in the pallet Config trait, I believe and therefore would require unnecessary boilerplate on consuming runtimes. Also

FYI and back to this: No, I don't think this is true. You can add this new `boolean` to the parameters of the `ExtBuilder` while not having it in the `Config` anywhere. If you want me to demonstrate this further, perhaps give me write permission to your fork's current branch so I can push an example there.

**Comment by ropottnik** [comment]
> > However, I didn't add the case distinction from the PR you referenced because it would have required adding a MockWeightInfo boolean type in the pallet Config trait, I believe and therefore would require unnecessary boilerplate on consuming runtimes. Also
> 
> FYI and back to this: No, I don't think this is true. You can add this new `boolean` to the parameters of the `ExtBuilder` while not having it in the `Config` anywhere. If you want me to demonstrate this further, perhaps give me write permission to your fork's current branch so I can push an example there.

sure, that would be most welcome! I saw that you can add these properties to the ExtBuilder. I found the idea weird, however, to make the `WeightInfo` depend on a type that is only available in a test environment. But I guess that I'm misunderstanding something :D

File: frame/elections-phragmen/src/lib.rs [hunk 8]
```diff
@@ -653,9 +657,13 @@ decl_module! {
...
 	/// The deposit value of `count` votes.
-	fn deposit_of(count: usize) -> BalanceOf<T> {
-		T::VotingBondBase::get().saturating_add(
-			T::VotingBondFactor::get().saturating_mul((count as u32).into())
+	fn deposit_of(count: u32) -> BalanceOf<T> {
+		T::VotingBondStorageBase::get().saturating_add(
+			T::VotingBondStorageFactor::get().saturating_mul(count.into())
+		).saturating_add(
+			T::VotingBondWeightFactor::get().saturating_mul(
+				T::WeightInfo::election_phragmen(0, 1, count).saturated_into()
```
Comment (kianenigma) [inline frame/elections-phragmen/src/lib.rs:665]: note that this was wrong before: additional weight to reserve is the weight for 0 candidate, 1 additional voter, and `count` votes (edges).

Comment (ropottnik) [reply frame/elections-phragmen/src/lib.rs:665]: ah, thanks!

File: frame/elections-phragmen/src/lib.rs [hunk 9]
```diff
@@ -1258,16 +1258,10 @@ mod tests {
...
-		pub fn voter_bond_storage_factor(self, bond: u64) -> Self {
-			VOTING_BOND_STORAGE_FACTOR.with(|v| *v.borrow_mut() = bond);
-			self
-		}
-		pub fn voter_bond_weight_factor(self, bond: u64) -> Self {
-			VOTING_BOND_WEIGHT_FACTOR.with(|v| *v.borrow_mut() = bond);
+		pub fn voter_bonds(self, base: u64, storage: u64, weight: u64) -> Self {
+			VOTING_BOND_STORAGE_BASE.with(|v| *v.borrow_mut() = base);
+			VOTING_BOND_STORAGE_FACTOR.with(|v| *v.borrow_mut() = storage);
+			VOTING_BOND_WEIGHT_FACTOR.with(|v| *v.borrow_mut() = weight);
```
Comment (ropottnik) [inline frame/elections-phragmen/src/lib.rs:1264]: 👍

File: frame/elections-phragmen/src/lib.rs (hunk 8 above)
Comment (thiolliere) [inline frame/elections-phragmen/src/lib.rs:665]: it is a bit unual to me to pay some weight with some deposit.

But I think the number of vote is limited thus we can't use too much weight anyway isn't it ?

File: frame/elections-phragmen/src/lib.rs [hunk 10]
```diff
@@ -2458,10 +2551,8 @@ mod tests {
 					..
 				}
 			);
-			matches!(
-				unwrapped_error.post_info.actual_weight,
-				Some(x) if x < <Test as frame_system::Config>::BlockWeights::get().max_block
```
Comment (thiolliere) [inline frame/elections-phragmen/src/lib.rs:2463]: why did you remove this matches ? you could actually keep them ?

Also we need to assert on these matches otherwise they are useless and don't test anything.

**Review by thiolliere — APPROVED** [review]
* code is good to me.
* The fact that we pay weight by deposit is a bit weird but should be fine as other constraint should ensure elect_phragmen doesn't take too much weight.
* test needs to fix these matches but it is not introduced by the PR

File: frame/elections-phragmen/src/lib.rs (hunk 10 above)
Comment (kianenigma) [reply frame/elections-phragmen/src/lib.rs:2463]: I added an assertion to the exact value. Checking that it is less than max block is no longer needed.

File: frame/elections-phragmen/src/lib.rs (hunk 8 above)
Comment (kianenigma) [reply frame/elections-phragmen/src/lib.rs:665]: Yes the deposit amount is limited. 

Indeed weight deposit is a new concept, but I think it makes sense. While you are imposing some computation on chain, you should at least reserve some stake for it.

File: frame/elections-phragmen/src/lib.rs (hunk 10 above)
Comment (thiolliere) [reply frame/elections-phragmen/src/lib.rs:2463]: ok, also the other matches should be asserted https://github.com/paritytech/substrate/pull/8211

**Comment by gnunicorn** [comment]
anyone working on this?

**Comment by kianenigma** [comment]
The audit has been delayed a lot here.

**Comment by kianenigma** [comment]
I did a cost analysis of the council stuff in polkadot recently and we don't really need this at this point, so for now I will close this. sorry for the confusion @ropottnik
